File: sample-py/src/householder.py
# -*- coding: utf-8 -*-

from typing import List

from math import sqrt

__all__ = ["householder"]

EPS = 1.0E-8


def householder(xs: List[float], num: int) -> bool:
    u  = [0.0 for i in range(num)]
    d  = [0.0 for i in range(num)]
    ds = [0.0 for i in range(num)]

    for i in range(num - 2):
        #  変換行列 H の構築
        if create_u(u, xs, i, num) is True:
            continue

        #  similarity transform
        init_d(d, xs, u, i, num)
        update_d(d, u, i + 1, num)

        init_ds(ds, xs, u, i, num)
        update_d(ds, u, i + 1, num)

        update_hessenberg(xs, u, d, ds, num)
    return False


def create_u(u: List[float], xs: List[float], k: int, num: int) -> bool:
    init_u(u, xs, k, num)

    total = cal_sum(u, k, num)
    if abs(u[k + 1]) < EPS:
        return True

    sigma = sqrt(total) * u[k + 1] / abs(u[k + 1])
    u[k + 1] += sigma
    norm = sqrt(2.0 * sigma * u[k + 1])
    for i in range(k + 1, num):
        u[i] /= norm
    return False


def init_u(ys: List[float], xs: List[float], k: int, num: int) -> None:
    for i in range(k + 1):
        ys[i] = 0.0
    for i in range(k + 1, num):
        ys[i] = xs[num * i + k]


def cal_sum(xs: List[float], begin: int, num: int) -> float:
    return sum(i * i for i in xs[begin + 1: num])
    # total = 0.0
    # for (int i = begin + 1; i < num; i++) {
    #     total += xs[i] * xs[i]
    # return total


def init_d(d: List[float], xs: List[float], u: List[float], k: int, length: int) -> None:
    for i in range(length):
        d[i] = 0.0
        for j in range(k + 1, length):
            d[i] += xs[length * i + j] * u[j]


def init_ds(ds: List[float], xs: List[float], u: List[float], k: int, length: int) -> None:
    for i in range(length):
        ds[i] = 0.0
        for j in range(k + 1, length):
            ds[i] += xs[length * j + i] * u[j]


def product_sum(xs: List[float], ys: List[float], begin: int, end: int) -> float:
    return sum(i * j for i, j in zip(xs[begin: end], ys[begin: end]))


def update_d(ys: List[float], xs: List[float], begin: int, end: int) -> None:
    ud  = product_sum(xs, ys, begin, end)
    for i in range(end):
        ys[i] = 2.0 * (ys[i] - ud * xs[i])


def update_hessenberg(
        xs: List[float],
        u:  List[float],
        d:  List[float],
        ds: List[float],
        num: int
) -> None:
    for i in range(num):
        for j in range(num):
            xs[num * i + j] -= u[i] * ds[j] + d[i] * u[j]

File: sample1/householder.c
#include "householder.h"

#include <stdio.h>
#include <stdlib.h>
#include <math.h>


#define EPS (1.0E-8)


bool check_calloc_error(double u[], double d[], double ds[]);

static void init_u(double *, const double *, const int k, const int n);
static double cal_sum(const double xs[], const int begin, const int num);
static bool create_u(double u[], const double xs[], const int k, const int num);
static void init_d(double d[], const double xs[], const double u[], const int k, const int length);
static void init_ds(double ds[], const double xs[], const double u[], const int k, const int length);
static double product_sum(const double xs[], const double ys[], const int begin, const int end);
static void update_d(double ys[], const double xs[], const int begin, const int end);
static void update_hessenberg(double [], const double [], const double [], const double [], const int);


/*
 * to upper hessenberg matrix with householder
 */
bool householder(double xs[], const int num){
  double *u = (double *)calloc(num, sizeof(double));
  double *d = (double *)calloc(num, sizeof(double));
  double *ds =(double *)calloc(num, sizeof(double));
  if (check_calloc_error(u, d, ds) == true) {
    return true;
  }

  for (int k = 0; k <= num - 3; k++) {
    //  変換行列 H の構築
    if (create_u(u, xs, k, num) == true) {
      continue;
    }

    //  similarity transform
    init_d(d, xs, u, k, num);
    update_d(d, u, k + 1, num);

    init_ds(ds, xs, u, k, num);
    update_d(ds, u, k + 1, num);

    update_hessenberg(xs, u, d, ds, num);
  }
  return false;
}

bool check_calloc_error(double u[], double d[], double ds[])
{
  if (u == NULL) {
    puts("u is calloc error");
    return true;
  }
  if (d == NULL) {
    puts("d is calloc error");
    return true;
  }
  if (ds == NULL) {
    puts("ds is calloc error");
    return true;
  }
  return false;
}


static void init_u(double ys[], const double xs[], const int k, const int num)
{
 for (int i = 0; i <= k; i++) {
   ys[i] = 0.0;
 }
 for (int i = k + 1; i < num; i++) {
   ys[i] = xs[num * i + k];
 }
}


static bool create_u(double u[], const double xs[], const int k, const int num)
{
  init_u(u, xs, k, num);

  double sum = cal_sum(u, k, num);
  if (fabs(u[k+1]) < EPS) {
    return true;
  }

  const double sigma = sqrt(sum) * u[k + 1] / fabs(u[k + 1]);
  u[k+1] += sigma;
  const double v_norm = sqrt(2.0 * sigma * u[k + 1]);
  for (int i = k + 1; i < num; i++) {
    u[i] /= v_norm;
  }
  return false;
}


static double cal_sum(const double xs[], const int begin, const int num)
{
  double sum = 0.0;
  for (int i = begin + 1; i < num; i++) {
    sum += xs[i] * xs[i];
  }
  return sum;
}


static void init_d(double d[], const double xs[], const double u[], const int k, const int length)
{
  for (int i = 0; i < length; i++) {
    d[i] = 0.0;
    for (int j = k + 1; j <= length - 1; j++) {
      d[i] += xs[length * i + j] * u[j];
    }
  }
}


static void init_ds(double ds[], const double xs[], const double u[], const int k, const int length)
{
  for (int i = 0; i < length; i++) {
    ds[i] = 0.0;
    for (int j = k + 1; j <= length - 1; j++) {
      ds[i] += xs[length * j + i] * u[j];
    }
  }
}


static double product_sum(const double xs[], const double ys[], const int begin, const int end)
{
  double total = 0.0;
  // for (int i = k + 1; i < end; i++) {
  for (int i = begin; i < end; i++) {
     total += xs[i] * ys[i];
  }
  return total;
}

static void update_d(double ys[], const double xs[], const int begin, const int end)
{
  const double ud  = product_sum(xs, ys,  begin, end);
  for (int i = 0; i < end; i++) {
     ys[i] = 2.0 * (ys[i] - ud * xs[i]);
  }
}

static void update_hessenberg(double xs[], const double u[], const double d[], const double ds[], const int num)
{
  for (int i = 0; i < num; i++) {
    for (int j = 0; j < num; j++) {
      xs[num * i + j] -= u[i] * ds[j] + d[i] * u[j];
    }
  }
}

